### PR TITLE
Implement slab containers and enhance mempools.

### PR DESCRIPTION
**Pull request by allensamuels2**
signed-off-by: Allen Samuels allen.samuels@sandisk.com

### CODE REVIEW

**Comment by dachary** [comment]
needs rebase :-)

**Comment by liewegas** [comment]
@rzarzynski care to resurrect this?

**Comment by rzarzynski** [comment]
@liewegas: sure!

**Comment by allensamuels2** [comment]
If you have any questions, I’m happy to answer them to the best of my memory….

File: src/include/slab_containers.h [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+  };
+  //
+  // Each slot has a pointer to it's containing slab PLUS either one Object OR if it's free exactly one pointer 
+  // in a per-slab freelist.
+  // Since this is raw memory, we don't want to declare something of type "T" to avoid
+  // accidental constructor/destructor calls.
+  //
+  struct slot_t {
+    slab_t *slab; // Pointer to my slab, NULL for slots on the stack (Technically, this could be an index but I'm lazy :))
+    enum { OBJECT_IN_POINTERS = (sizeof(T) + sizeof(void *) - 1) / sizeof(void *) };
```
Comment (rzarzynski) [inline src/include/slab_containers.h]: Maybe `std::aligned_storage` instead.

**Comment by allensamuels2** [comment]
If the underlying object has an alignment restriction greater than a pointer, this code will fail miserably. I don’t know if an assert to that effect can be done or not (maybe something like: assert(alignof(T) <= sizeof(void *));    ????

The usage of the OBJECT_IN_POINTERS is done because there is an implicit union of the template type <T> with a pointer (when the slot is not in use, a pointer to the free list is stored). The ENUM’s computation guarantees that if sizeof(T) is less than sizeof(void *) you are still reserving sufficient space. So you could just make this explicit with a union -> that would probably make the code a bit clearer.

**Comment by liewegas** [comment]
see @rzarzynski 's newer implementations of this

**Pull request closed without merging**